File: lib_rxdownloader/src/main/java/com/wuzhu/rx/downloader/Downloader.java
package com.wuzhu.rx.downloader;

/**
 * @author Hdq on 2021/3/10.
 */
public class Downloader {
    private static final String TAG = "----Downloader";
    private static volatile DownloadGroup defaultDownloadGroup;
    public static DownloadGroup defaultDownloadGroup(){
        if (defaultDownloadGroup == null) {
            synchronized (Downloader.class){
                if (defaultDownloadGroup == null) {
                    defaultDownloadGroup = new DownloadGroup(5);
                    defaultDownloadGroup.start();
                }
            }
        }
        return defaultDownloadGroup;
    }

    public static void stopDefaultDownloadGroup(){
        stopDownloadGroup(defaultDownloadGroup);
    }

    public static DownloadGroup newDownloadGroup(){
        return newDownloadGroup(5);
    }

    /**
     * 最大并发下载数量
     * @param limit
     * @return
     */
    public static DownloadGroup newDownloadGroup(int limit){
        return new DownloadGroup(limit);
    }

    public static void stopDownloadGroup(DownloadGroup downloadGroup){
        if (downloadGroup != null) {
            downloadGroup.stopGroup();
        }
    }

}

File: lib_rxdownloader/src/main/java/com/wuzhu/rx/downloader/Downloader.kt
package com.wuzhu.rx.downloader

import android.util.Log
import androidx.annotation.Keep
import com.wuzhu.rx.downloader.exceptions.DownloadException
import com.wuzhu.rx.downloader.utils.FilePathUtils
import okhttp3.OkHttpClient
import java.io.File
import java.util.concurrent.ConcurrentHashMap

/**
 * @author Hdq on 2021/3/10.
 */
@Keep
object Downloader {
    private const val TAG = "----Downloader"
    private val downloadGroups = ConcurrentHashMap<String, DownloadGroup>()
    var client: OkHttpClient? = null

    @JvmStatic
    fun init(client: OkHttpClient?) {
        Log.d(TAG, "init: ")
        this.client = client
    }

    /**
     * 最大并发下载数量
     *
     * @param limit
     * @return
     */
    @JvmOverloads
    @JvmStatic
    @Synchronized
    fun createDownloadGroup(key: String, limit: Int = 5): DownloadGroup {
        if (downloadGroups.containsKey(key)) {
            throw DownloadException("key重复：：以key=$key 创建的任务已经存在")
        }
        val downloadGroup = DownloadGroup(limit)
        downloadGroup.create(key, client)
        downloadGroups[key] = downloadGroup
        return downloadGroup
    }

    @JvmStatic
    fun findDownloadGroup(key: String): DownloadGroup? {
        return downloadGroups[key]
    }

    @JvmOverloads
    @JvmStatic
    @Synchronized
    fun findOrCreateDownloadGroup(key: String, limit: Int = 5): DownloadGroup {
        var downloadGroup = downloadGroups[key]
        downloadGroup ?: run { downloadGroup = createDownloadGroup(key, limit) }
        return downloadGroup!!
    }

    @Suppress("unused")
    @JvmStatic
    fun isExistDownloadGroup(key: String): Boolean {
        return downloadGroups.containsKey(key)
    }

    @Suppress("unused")
    @JvmStatic
    fun getDownloadGroups() = downloadGroups as Map<String, DownloadGroup>

    @JvmOverloads
    @JvmStatic
    @Synchronized
    fun startDownloadGroup(key: String, limit: Int = 5): DownloadGroup {
        val downloadGroup = findOrCreateDownloadGroup(key, limit)
        downloadGroup.start()
        return downloadGroup
    }

    @JvmStatic
    @Synchronized
    fun stopDownloadGroup(key: String) {
        findDownloadGroup(key)?.stop()
    }


    @JvmStatic
    @Synchronized
    fun destroyAllGroup() {
        downloadGroups.forEach {
            it.value.destroy()
        }
        downloadGroups.clear()
    }

    @JvmStatic
    @Synchronized
    fun destroyGroup(key: String) {
        findDownloadGroup(key)?.destroy()
        downloadGroups.remove(key)
    }

    @JvmStatic
    fun deleteTempFile(localFile: String, downloadingMd5: String?): Boolean {
        val tempFile = File(FilePathUtils.getTempFileName(localFile, downloadingMd5))
        if (tempFile.exists()) {
            return tempFile.delete()
        }
        return false
    }

}